File: Mini-Games/readability/readability-py/readability.py
from cs50 import get_string


def main():
    # Get the Text from the user
    text = get_string("Text: ")
    while not check_correctness(text):
        text = get_string("Text: ")

    # Store the variables
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # Compute the index and printing it
    Coleman_Liau_index(letters, words, sentences)


# Checking the Correctness
def check_correctness(text):
    for i in range(len(text) - 1):
        # Elimination of consecutive spaces
        if (text[i] == ' ') and (text[i + 1] == ' '):
            print("It looks like you used consecutive spaces in your text.")
            return False
        # Elimination of wrong Punctuation
        elif (text[i] == '!' and (i == 0 or i == 1)) or (text[i] == '?' and (i == 0 or i == 1)) or (text[i] == '.' and (i == 0 or i == 1)) or (text[i] == ',' and (i == 0 or i == 1)):
            print("It looks like you used Punctuation in the beginning of your text.")
            return False
    return True


# Calculating the number of Letters
def count_letters(text):
    sum = 0
    for i in range(len(text)):
        if (text[i] >= 'A' and text[i] <= 'Z') or (text[i] >= 'a' and text[i] <= 'z'):
            sum += 1
    return sum


# Calculating the number of Words
def count_words(text):
    sum = 0
    for i in range(len(text)):
        if text[i] == ' ':
            sum += 1
    return sum + 1


# Calculating the number of Sentences
def count_sentences(text):
    sum = 0
    n = len(text)
    if text[n - 1] == '!' or text[n - 1] == '?' or text[n - 1] == '.':
        sum += 1
    for i in range(len(text)):
        # Exceptional cases filtering
        if text[i] == '.' and text[i - 3] != ' ' and text[i - 4] != ' ' and i != n - 1 and i != 2 and i != 3:
            sum += 1
        elif (text[i] == '.' and text[i - 3] == ' ' and text[i - 2].isupper() == False and i != n - 1 and i != 2 and i != 3) or (text[i] == '.' and text[i - 4] == ' ' and text[i - 3].isupper() == False and i != n - 1 and i != 2 and i != 3):
            sum += 1
        elif i <= n - 3:
            if text[i] == '!' and text[i + 2].isupper() == True:
                sum += 1
            elif text[i] == '?' and text[i + 2].isupper() == True:
                sum += 1
    return sum


# Computing the index and printing it
def Coleman_Liau_index(letters, words, sentences):
    # Parameters calculation
    l = (float(letters) / float(words)) * 100
    s = (float(sentences) / float(words)) * 100
    # Index calculation
    g = round(0.0588 * l - 0.296 * s - 15.8)
    # Index Printing
    if g < 1:
        print("Before Grade 1")
    elif g >= 1 and g <= 15:
        print(f"Grade {g}")
    elif g >= 16:
        print("Grade 16+")


if __name__ == "__main__":
    main()

File: Mini-Games/readability/readability-c/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

// Declaring The Fonctions
bool check_correctness(string text);
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
void Coleman_Liau_index(int letters, int words, int sentences);

int main(void)
{
    // Get the Text from the user
    string text = get_string("Text: ");
    while (check_correctness(text) == false)
    {
        text = get_string("Text: ");
    }
    // Store the variables
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences =  count_sentences(text);
    // Compute the index and printing it
    Coleman_Liau_index(letters, words, sentences);
}

// Checking the Correctness
bool check_correctness(string text)
{
    for (int i = 0, n = strlen(text); i < n - 1; i++)
    {
        // Elimination of consecutive spaces
        if (text[i] == ' ' && text[i + 1] == ' ')
        {
            printf("It looks like you used consecutive spaces in your text.\n");
            return false;
        }
        // Elimination of wrong Punctuation
        else if ((text[i] == '!' && (i == 0 || i == 1)) || (text[i] == '?' && (i == 0 || i == 1)) || (text[i] == '.' && (i == 0 || i == 1))
                 || (text[i] == ',' && (i == 0 || i == 1)))
        {
            printf("It looks like you used Punctuation in the beginning of your text.\n");
            return false;
        }
    }
    return true;
}

// Finding the number of Letters
int count_letters(string text)
{
    int sum = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if ((text[i] >= 'A' && text[i] <= 'Z') || (text[i] >= 'a' && text[i] <= 'z'))
        {
            sum++;
        }
    }
    return sum;
}

// Finding the number of Words
int count_words(string text)
{
    int sum = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            sum++;
        }
    }
    return sum + 1;
}

// Finding the number of Sentences
int count_sentences(string text)
{
    int sum = 0;
    int n = strlen(text);
    if (text[n - 1] == '!' || text[n - 1] == '?' || text[n - 1] == '.')
    {
        sum++;
    }
    for (int i = 0; i < n; i++)
    {
        // Exceptional cases filtering
        if (text[i] == '.' && text[i - 3] != ' ' && text[i - 4] != ' ' && i != n - 1 && i != 2 && i != 3)
        {
            sum++;
        }
        else if ((text[i] == '.' && text[i - 3] == ' ' && isupper(text[i - 2]) == 0 && i != n - 1 && i != 2 && i != 3) || (text[i] == '.'
                 && text[i - 4] == ' ' && isupper(text[i - 3]) == 0 && i != n - 1 && i != 2 && i != 3))
        {
            sum++;
        }
        else if (text[i] == '!' && isupper(text[i + 2]) != 0 && i != n - 1)
        {
            sum++;
        }
        else if (text[i] == '?' && isupper(text[i + 2]) != 0 && i != n - 1)
        {
            sum++;
        }
    }
    return sum;
}

// Computing the index and printing it
void Coleman_Liau_index(int letters, int words, int sentences)
{
    // Parameters calculation
    float l = ((float) letters / (float) words) * 100;
    float s = ((float) sentences / (float) words) * 100;
    // Index calculation
    int g = round(0.0588 * l - 0.296 * s - 15.8);
    // Index Printing
    if (g < 1)
    {
        printf("Before Grade 1\n");
    }
    if (g >= 1 && g <= 15)
    {
        printf("Grade %i\n", g);
    }
    if (g >= 16)
    {
        printf("Grade 16+\n");
    }
}